src/Agent.Listener/Environment.cs

```diff
@@ -0,0 +1,338 @@
+using Microsoft.VisualStudio.Services.Agent.Util;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Listener
+{
+    [ServiceLocator(Default = typeof(Environment))]
+    public interface IEnvironment : IAgentService
+    {
+        Task<Dictionary<string, string>> GetCapabilities(string agentName, CancellationToken token);
+
+        void EnsureEnvFile();
+
+        IDictionary<string, string> GetEnv();
+    }
+
+    struct Capability
```

Comment:
Every time I see a struct, I think assume the author intends for the type to be optimized for the call stack. There may other reasons to use a struct, I don't know. I think very recent .Net versions have removed some of the restrictions that struct has (that class doesn't have).